src/core/qgsabstractdatabaseproviderconnection.h

```diff
@@ -139,14 +183,26 @@ class CORE_EXPORT QgsAbstractDatabaseProviderConnection : public QgsAbstractProv
 ///@cond private
 
         /**
-         * The QueryResultIterator struct is an abstract interface for provider query results iterators.
+         * The QueryResultIterator struct is an abstract interface for provider query result iterators.
          * Providers must implement their own concrete iterator over query results.
+         *
+         * \note This struct is thread safe.
          */
-        struct QueryResultIterator SIP_SKIP
+        struct CORE_EXPORT QueryResultIterator SIP_SKIP
         {
-          virtual QVariantList nextRow() = 0;
-          virtual bool hasNextRow() const = 0;
-          virtual ~QueryResultIterator() = default;
+            QVariantList nextRow();
+            bool hasNextRow() const;
+            qlonglong fetchedRowCount();
+            virtual ~QueryResultIterator() = default;
+            QList<QList<QVariant> > rows() const;
+
+          private:
+
+            virtual QVariantList nextRowPrivate() = 0;
+            virtual bool hasNextRowPrivate() const = 0;
+            mutable QList<QList<QVariant>> mRows;
```

Comment:
If we're storing all the fetched results in a list in memory, doesn't this negate the benefit of having an iterator?

Reply:
Memory-wise, yes. There is still an advantage in using the iterator that is that the fetching is lazy. 

This is the only way I've found to make it work from multiple threads and with `rows()` without re-running the query every time, which we don't want to happen because the query might have side-effects.

Maybe the root issue is that this implementation started with `rows()`, then I added the iterator later. So, if you call `rows()` multiple times you still fetch the rows only once, additionally you can also call `rows()` and use the iterator at the same time from multiple threads without issues. `at()` is also supported.

Reply:
@elpaso Good point regarding the model requiring the storage of all rows, but I think ideally it should be model's responsibility to store the list of rows instead of the iterator's. Then the iterator could be a "pure iterator", as memory-friendly as possible and could be used in other contexts (ie outside of the model) without inheriting the storage requirement of the model.

Reply:
And to add to that -- we could make the QueryResult::rows() method still iterate over all rows, but remove the mRows member and just build a list in the rows() method internally and return that. So `rows` will still be slow to call and subject to potential memory limitations, but nextRow() will be as light as possible..

Reply:
The problem is that we cannot run `rows()` two times without re-running the query (with possible side-effects), this is why I was storing the results inside the class.

But I'm ok with turning it into a real iterator over the rows, but we'll need to drop `rows()` or just make it return an empty set when called the second time.